File: benchmarks/julia/benchmark_projected_apply.jl
# Benchmark isolated ITensors.ProjMPO local apply calls.
#
# Run:
#   julia --project=benchmarks/julia benchmarks/julia/benchmark_projected_apply.jl
#
# Optional args:
#   julia --project=benchmarks/julia benchmarks/julia/benchmark_projected_apply.jl <N> <state_bond_dim> <operator_bond_dim> <repeats> <step_index>
#
# For a one-thread comparison with the Rust example:
#   BLAS_NUM_THREADS=1 julia --project=benchmarks/julia benchmarks/julia/benchmark_projected_apply.jl 38 32 32 3 0

import Pkg

Pkg.activate(@__DIR__)
Pkg.instantiate()

using ITensors
using LinearAlgebra
using Printf
using Random

ITensors.disable_warn_order()

function parse_positive_int_arg(args::Vector{String}, index::Int, default::Int, name::String)::Int
    value = index <= length(args) ? parse(Int, args[index]) : default
    value > 0 || error("$name must be greater than zero")
    return value
end

function parse_nonnegative_int_arg(args::Vector{String}, index::Int, default::Int, name::String)::Int
    value = index <= length(args) ? parse(Int, args[index]) : default
    value >= 0 || error("$name must be nonnegative")
    return value
end

function maybe_set_blas_threads_from_env!()
    haskey(ENV, "BLAS_NUM_THREADS") || return
    nthreads = parse(Int, ENV["BLAS_NUM_THREADS"])
    nthreads > 0 || error("BLAS_NUM_THREADS must be greater than zero")
    BLAS.set_num_threads(nthreads)
end

function state_indices(
    site::Int,
    nsites::Int,
    acted_sites::Vector{Index{Int64}},
    spectator_sites::Vector{Index{Int64}},
    state_links::Vector{Index{Int64}},
)
    sites = Index{Int64}[spectator_sites[site], acted_sites[site]]
    if nsites == 1
        return sites
    elseif site == 1
        return vcat(sites, Index{Int64}[state_links[site]])
    elseif site == nsites
        return vcat(Index{Int64}[state_links[site - 1]], sites)
    else
        return vcat(Index{Int64}[state_links[site - 1]], sites, Index{Int64}[state_links[site]])
    end
end

function operator_indices(site::Int, nsites::Int, operator_links::Vector{Index{Int64}})
    if nsites == 1
        return Index{Int64}[]
    elseif site == 1
        return Index{Int64}[operator_links[site]]
    elseif site == nsites
        return Index{Int64}[operator_links[site - 1]]
    else
        return Index{Int64}[operator_links[site - 1], operator_links[site]]
    end
end

function make_state_mps(
    rng::AbstractRNG,
    nsites::Int,
    state_bond_dim::Int,
    acted_sites::Vector{Index{Int64}},
    spectator_sites::Vector{Index{Int64}},
)::MPS
    state_links = [Index(state_bond_dim, "Link,psi,l=$site") for site in 1:(nsites - 1)]
    tensors = Vector{ITensor}(undef, nsites)
    for site in 1:nsites
        indices = state_indices(site, nsites, acted_sites, spectator_sites, state_links)
        tensors[site] = random_itensor(rng, indices...)
    end
    return MPS(tensors)
end

function make_operator_mpo(
    rng::AbstractRNG,
    nsites::Int,
    operator_bond_dim::Int,
    acted_sites::Vector{Index{Int64}},
    spectator_sites::Vector{Index{Int64}},
)::MPO
    operator_links = [Index(operator_bond_dim, "Link,H,l=$site") for site in 1:(nsites - 1)]
    tensors = Vector{ITensor}(undef, nsites)
    for site in 1:nsites
        core_indices = vcat(
            operator_indices(site, nsites, operator_links),
            Index{Int64}[acted_sites[site], prime(acted_sites[site])],
        )
        core = random_itensor(rng, core_indices...)
        spectator_identity = delta(spectator_sites[site], prime(spectator_sites[site]))
        tensors[site] = core * spectator_identity
    end
    return MPO(tensors)
end

function two_site_sweep_positions(nsites::Int, center_pos::Int)::Vector{Int}
    positions = collect(center_pos:(nsites - 1))
    append!(positions, collect((center_pos - 1):-1:1))
    return positions
end

function elapsed_seconds(f)::Tuple{Float64, Any}
    start = time_ns()
    result = f()
    return ((time_ns() - start) / 1.0e9, result)
end

function summarize(label::String, times::Vector{Float64})
    mean = sum(times) / length(times)
    min_time = minimum(times)
    max_time = maximum(times)
    @printf(
        "%s: mean=%.3f ms min=%.3f ms max=%.3f ms n=%d\n",
        label,
        mean * 1000.0,
        min_time * 1000.0,
        max_time * 1000.0,
        length(times),
    )
end

function main(args::Vector{String})
    nsites = parse_positive_int_arg(args, 1, 38, "N")
    state_bond_dim = parse_positive_int_arg(args, 2, 8, "state_bond_dim")
    operator_bond_dim = parse_positive_int_arg(args, 3, 8, "operator_bond_dim")
    repeats = parse_positive_int_arg(args, 4, 20, "repeats")
    step_index = parse_nonnegative_int_arg(args, 5, 0, "step_index")

    nsites >= 2 || error("N must be at least 2 for a two-site local step")
    maybe_set_blas_threads_from_env!()

    phys_dim = 2
    seed = 20260518
    rng = MersenneTwister(seed)

    acted_sites = [Index(phys_dim, "s=$site") for site in 1:nsites]
    spectator_sites = [Index(phys_dim, "q=$site") for site in 1:nsites]

    psi = make_state_mps(rng, nsites, state_bond_dim, acted_sites, spectator_sites)
    H = make_operator_mpo(rng, nsites, operator_bond_dim, acted_sites, spectator_sites)

    center_pos = nsites ÷ 2 + 1
    positions = two_site_sweep_positions(nsites, center_pos)
    local_pos = positions[mod(step_index, length(positions)) + 1]
    phi = psi[local_pos] * psi[local_pos + 1]

    # Compile the relevant ITensor/ProjMPO code before taking timings.
    warmup_projected = ITensors.ProjMPO(H)
    position!(warmup_projected, psi, local_pos)
    warmup_out = warmup_projected(phi)
    GC.@preserve warmup_out nothing
    GC.gc()

    println("=== ProjMPO local apply benchmark ===")
    println("N = $nsites")
    println("phys_dim = $phys_dim")
    println("state_bond_dim = $state_bond_dim")
    println("operator_bond_dim = $operator_bond_dim")
    println("repeats = $repeats")
    println("threads = $(Threads.nthreads())")
    println("blas_threads = $(BLAS.get_num_threads())")
    println("center_pos = $center_pos")
    println("step_index = $(mod(step_index, length(positions)))")
    println("local_sites = ($local_pos, $(local_pos + 1))")
    println("local_dims = $(dim.(inds(phi)))")
    println()

    projected_ref = Ref{Any}()
    cold_time, cold_out = elapsed_seconds() do
        projected = ITensors.ProjMPO(H)
        position!(projected, psi, local_pos)
        out = projected(phi)
        projected_ref[] = projected
        return out
    end
    @printf(
        "cold apply (environment build + one apply): %.3f ms, output_rank=%d\n",
        cold_time * 1000.0,
        order(cold_out),
    )

    projected = projected_ref[]
    warm_times = Float64[]
    sizehint!(warm_times, repeats)
    warm_order_sum = 0
    for _ in 1:repeats
        t, out = elapsed_seconds() do
            return projected(phi)
        end
        warm_order_sum += order(out)
        push!(warm_times, t)
    end
    summarize("warm apply (environment cache hot)", warm_times)

    cold_times = Float64[]
    sizehint!(cold_times, repeats)
    cold_order_sum = 0
    for _ in 1:repeats
        t, out = elapsed_seconds() do
            projected_cold = ITensors.ProjMPO(H)
            position!(projected_cold, psi, local_pos)
            return projected_cold(phi)
        end
        cold_order_sum += order(out)
        push!(cold_times, t)
    end
    summarize("cold apply repeated (fresh environment cache)", cold_times)

    GC.@preserve warm_order_sum cold_order_sum nothing
    return nothing
end

main(ARGS)

File: benchmarks/rust/benchmark_projected_apply.rs
// Benchmark isolated `ProjectedOperator::apply` calls for mapped MPO local solves.
//
// Run:
//   cargo run -p tensor4all-treetn --example benchmark_projected_apply --release
//
// Optional args:
//   cargo run -p tensor4all-treetn --example benchmark_projected_apply --release -- <N> <state_bond_dim> <operator_bond_dim> <repeats> <step_index>

use std::collections::{HashMap, HashSet};
use std::hint::black_box;
use std::time::{Duration, Instant};

use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use tensor4all_core::{
    index::{DynId, Index},
    DynIndex, TensorDynLen, TensorIndex,
};
use tensor4all_treetn::{IndexMapping, LocalUpdateSweepPlan, ProjectedOperator, TreeTN};

fn make_node_name(i: usize) -> String {
    format!("site{i}")
}

fn unique_dyn_index(used: &mut HashSet<DynId>, dim: usize, rng: &mut StdRng) -> DynIndex {
    loop {
        let id = DynId(rng.random());
        if used.insert(id) {
            return Index::new(id, dim);
        }
    }
}

fn chain_node_indices(n: usize, i: usize, bonds: &[DynIndex], sites: &[DynIndex]) -> Vec<DynIndex> {
    if n == 1 {
        vec![sites[i].clone()]
    } else if i == 0 {
        vec![sites[i].clone(), bonds[i].clone()]
    } else if i + 1 == n {
        vec![bonds[i - 1].clone(), sites[i].clone()]
    } else {
        vec![bonds[i - 1].clone(), sites[i].clone(), bonds[i].clone()]
    }
}

fn create_state_chain(
    n: usize,
    state_bond_dim: usize,
    acted_sites: &[DynIndex],
    spectator_sites: &[DynIndex],
    used_ids: &mut HashSet<DynId>,
    rng: &mut StdRng,
) -> anyhow::Result<TreeTN<TensorDynLen, String>> {
    let mut tree = TreeTN::<TensorDynLen, String>::new();
    let bonds: Vec<_> = (0..n.saturating_sub(1))
        .map(|_| unique_dyn_index(used_ids, state_bond_dim, rng))
        .collect();

    let mut nodes = Vec::with_capacity(n);
    for i in 0..n {
        let mut indices = chain_node_indices(n, i, &bonds, acted_sites);
        indices.insert(0, spectator_sites[i].clone());
        let tensor = TensorDynLen::random::<f64, _>(rng, indices)?;
        let node = tree.add_tensor(make_node_name(i), tensor)?;
        nodes.push(node);
    }

    for i in 0..n.saturating_sub(1) {
        tree.connect(nodes[i], &bonds[i], nodes[i + 1], &bonds[i])?;
    }

    Ok(tree)
}

#[allow(clippy::type_complexity)]
fn create_operator_chain(
    n: usize,
    phys_dim: usize,
    operator_bond_dim: usize,
    acted_sites: &[DynIndex],
    used_ids: &mut HashSet<DynId>,
    rng: &mut StdRng,
) -> anyhow::Result<(
    TreeTN<TensorDynLen, String>,
    HashMap<String, IndexMapping<DynIndex>>,
    HashMap<String, IndexMapping<DynIndex>>,
)> {
    let mut tree = TreeTN::<TensorDynLen, String>::new();
    let bonds: Vec<_> = (0..n.saturating_sub(1))
        .map(|_| unique_dyn_index(used_ids, operator_bond_dim, rng))
        .collect();
    let s_in: Vec<_> = (0..n)
        .map(|_| unique_dyn_index(used_ids, phys_dim, rng))
        .collect();
    let s_out: Vec<_> = (0..n)
        .map(|_| unique_dyn_index(used_ids, phys_dim, rng))
        .collect();

    let mut input_mapping = HashMap::new();
    let mut output_mapping = HashMap::new();
    let mut nodes = Vec::with_capacity(n);
    for i in 0..n {
        let sites = vec![s_out[i].clone(), s_in[i].clone()];
        let mut indices = if n == 1 {
            sites
        } else if i == 0 {
            vec![s_out[i].clone(), s_in[i].clone(), bonds[i].clone()]
        } else if i + 1 == n {
            vec![bonds[i - 1].clone(), s_out[i].clone(), s_in[i].clone()]
        } else {
            vec![
                bonds[i - 1].clone(),
                s_out[i].clone(),
                s_in[i].clone(),
                bonds[i].clone(),
            ]
        };
        indices.shrink_to_fit();
        let tensor = TensorDynLen::random::<f64, _>(rng, indices)?;
        let name = make_node_name(i);
        let node = tree.add_tensor(name.clone(), tensor)?;
        nodes.push(node);

        input_mapping.insert(
            name.clone(),
            IndexMapping {
                true_index: acted_sites[i].clone(),
                internal_index: s_in[i].clone(),
            },
        );
        output_mapping.insert(
            name,
            IndexMapping {
                true_index: acted_sites[i].clone(),
                internal_index: s_out[i].clone(),
            },
        );
    }

    for i in 0..n.saturating_sub(1) {
        tree.connect(nodes[i], &bonds[i], nodes[i + 1], &bonds[i])?;
    }

    Ok((tree, input_mapping, output_mapping))
}

fn summarize(label: &str, times: &[Duration]) {
    let secs: Vec<f64> = times.iter().map(Duration::as_secs_f64).collect();
    let mean = secs.iter().sum::<f64>() / secs.len() as f64;
    let min = secs.iter().copied().fold(f64::INFINITY, f64::min);
    let max = secs.iter().copied().fold(f64::NEG_INFINITY, f64::max);
    println!(
        "{label}: mean={:.3} ms min={:.3} ms max={:.3} ms n={}",
        mean * 1000.0,
        min * 1000.0,
        max * 1000.0,
        times.len()
    );
}

fn main() -> anyhow::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let n_sites: usize = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(38);
    let state_bond_dim: usize = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(8);
    let operator_bond_dim: usize = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(8);
    let repeats: usize = args.get(4).and_then(|s| s.parse().ok()).unwrap_or(20);
    let step_index: usize = args.get(5).and_then(|s| s.parse().ok()).unwrap_or(0);

    anyhow::ensure!(
        n_sites >= 2,
        "N must be at least 2 for a two-site local step"
    );
    anyhow::ensure!(repeats > 0, "repeats must be greater than zero");

    let phys_dim = 2usize;
    let seed = 20260518_u64;
    let mut used_ids = HashSet::<DynId>::new();
    let mut rng = StdRng::seed_from_u64(seed);

    let acted_sites: Vec<_> = (0..n_sites)
        .map(|_| unique_dyn_index(&mut used_ids, phys_dim, &mut rng))
        .collect();
    let spectator_sites: Vec<_> = (0..n_sites)
        .map(|_| unique_dyn_index(&mut used_ids, phys_dim, &mut rng))
        .collect();

    let state = create_state_chain(
        n_sites,
        state_bond_dim,
        &acted_sites,
        &spectator_sites,
        &mut used_ids,
        &mut rng,
    )?;
    let reference_state = state.sim_linkinds()?;
    let (operator, input_mapping, output_mapping) = create_operator_chain(
        n_sites,
        phys_dim,
        operator_bond_dim,
        &acted_sites,
        &mut used_ids,
        &mut rng,
    )?;

    let center = make_node_name(n_sites / 2);
    let plan = LocalUpdateSweepPlan::from_treetn(&state, &center, 2)
        .ok_or_else(|| anyhow::anyhow!("failed to build two-site sweep plan"))?;
    let step = plan
        .steps
        .get(step_index % plan.steps.len())
        .ok_or_else(|| anyhow::anyhow!("empty sweep plan"))?;
    let local_tensor = state.extract_subtree(&step.nodes)?.contract_to_tensor()?;

    println!("=== ProjectedOperator::apply benchmark ===");
    println!("N = {n_sites}");
    println!("phys_dim = {phys_dim}");
    println!("state_bond_dim = {state_bond_dim}");
    println!("operator_bond_dim = {operator_bond_dim}");
    println!("repeats = {repeats}");
    println!("center = {center}");
    println!("step_index = {}", step_index % plan.steps.len());
    println!("step_nodes = {:?}", step.nodes);
    println!("local_dims = {:?}", local_tensor.dims());
    println!();

    let mut projected_cold = ProjectedOperator::with_index_mappings(
        operator.clone(),
        input_mapping.clone(),
        output_mapping.clone(),
    );
    let cold_start = Instant::now();
    let cold_result = projected_cold.apply(
        black_box(&local_tensor),
        &step.nodes,
        &state,
        &reference_state,
        state.site_index_network(),
    )?;
    let cold = cold_start.elapsed();
    println!(
        "cold apply (environment build + one apply): {:.3} ms, output_rank={}",
        cold.as_secs_f64() * 1000.0,
        cold_result.external_indices().len()
    );

    let mut warm_times = Vec::with_capacity(repeats);
    for _ in 0..repeats {
        let start = Instant::now();
        let out = projected_cold.apply(
            black_box(&local_tensor),
            &step.nodes,
            &state,
            &reference_state,
            state.site_index_network(),
        )?;
        black_box(out);
        warm_times.push(start.elapsed());
    }
    summarize("warm apply (environment cache hot)", &warm_times);

    let mut cold_times = Vec::with_capacity(repeats);
    for _ in 0..repeats {
        let mut projected = ProjectedOperator::with_index_mappings(
            operator.clone(),
            input_mapping.clone(),
            output_mapping.clone(),
        );
        let start = Instant::now();
        let out = projected.apply(
            black_box(&local_tensor),
            &step.nodes,
            &state,
            &reference_state,
            state.site_index_network(),
        )?;
        black_box(out);
        cold_times.push(start.elapsed());
    }
    summarize("cold apply repeated (fresh environment cache)", &cold_times);

    Ok(())
}
